Issues with Discriminator Multitenant and inheritance

I stepped in a problem when using multi-tenancy in mode `DISCRIMINATOR` in sub classes with "non-multitenant" super classes. Stacktrace:
```
org.springframework.orm.hibernate5.HibernateSystemException: No such filter configured [tenantId]; nested exception is org.hibernate.HibernateException: No such filter configured [tenantId]
	at org.springframework.orm.hibernate5.SessionFactoryUtils.convertHibernateAccessException(SessionFactoryUtils.java:296)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.convertHibernateAccessException(GrailsHibernateTemplate.java:724)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:303)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:243)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:117)
	at org.grails.orm.hibernate.HibernateGormStaticApi.list(HibernateGormStaticApi.groovy:74)
	at org.grails.orm.hibernate.HibernateGormStaticApi.list(HibernateGormStaticApi.groovy:73)
	at org.grails.datastore.gorm.GormEntity$Trait$Helper.list(GormEntity.groovy:680)
	at ChildControllerSpec._tt___spock_feature_0_0_closure3(ChildControllerSpec.groovy:22)
	at groovy.lang.Closure.call(Closure.java:418)
	at groovy.lang.Closure.call(Closure.java:434)
	at grails.gorm.multitenancy.Tenants.withId_closure2(Tenants.groovy:250)
	at grails.gorm.multitenancy.Tenants$CurrentTenant.withTenant(Tenants.groovy:358)
	at grails.gorm.multitenancy.Tenants.withId(Tenants.groovy:236)
	at grails.gorm.multitenancy.Tenants.withId(Tenants.groovy:169)
	at ChildControllerSpec.$tt__$spock_feature_0_0(ChildControllerSpec.groovy:21)
	at ChildControllerSpec.should list some_closure1(ChildControllerSpec.groovy)
	at groovy.lang.Closure.call(Closure.java:418)
	at groovy.lang.Closure.call(Closure.java:434)
	at grails.gorm.transactions.GrailsTransactionTemplate$1.doInTransaction(GrailsTransactionTemplate.groovy:68)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at grails.gorm.transactions.GrailsTransactionTemplate.executeAndRollback(GrailsTransactionTemplate.groovy:65)
	at ChildControllerSpec.should list some(ChildControllerSpec.groovy)
Caused by: org.hibernate.HibernateException: No such filter configured [tenantId]
	at org.hibernate.internal.SessionFactoryImpl.getFilterDefinition(SessionFactoryImpl.java:1178)
	at org.hibernate.engine.spi.LoadQueryInfluencers.enableFilter(LoadQueryInfluencers.java:109)
	at org.hibernate.internal.SessionImpl.enableFilter(SessionImpl.java:2202)
	at org.grails.orm.hibernate.AbstractHibernateDatastore.enableMultiTenancyFilter(AbstractHibernateDatastore.java:383)
	at org.grails.orm.hibernate.multitenancy.MultiTenantEventListener.onApplicationEvent(MultiTenantEventListener.java:51)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:393)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:347)
	at org.grails.datastore.gorm.events.ConfigurableApplicationContextEventPublisher.publishEvent(ConfigurableApplicationContextEventPublisher.groovy:30)
	at org.grails.orm.hibernate.query.AbstractHibernateQuery.listForCriteria(AbstractHibernateQuery.java:716)
	at org.grails.orm.hibernate.HibernateGormStaticApi.list_closure1(HibernateGormStaticApi.groovy:88)
	at groovy.lang.Closure.call(Closure.java:418)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:299)
	... 20 more
```
I also made a project reproducing the error. The test triggers the exception:

[ex-multinenancy-inheritance.zip](https://github.com/grails/gorm-hibernate5/files/2116197/ex-multinenancy-inheritance.zip)


I tracked down the problem and seems like the class GrailsDomainBinder is only binding the tenant filter for root classes. There's any reason for that? There would be any problems binding the filters for subclasses?

I'm creating a fix pull request already, because we are already using the patched version and seems all good so far.

--- a/grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/cfg/GrailsDomainBinder.java
+++ b/grails-datastore-gorm-hibernate5/src/main/groovy/org/grails/orm/hibernate/cfg/GrailsDomainBinder.java
@@ -1408,22 +1408,44 @@ public class GrailsDomainBinder implements MetadataContributor {
             bindSubClasses(entity, root, mappings, sessionFactoryBeanName);
         }
 
-        if(entity.isMultiTenant()) {
-            TenantId tenantId = entity.getTenantId();
-            if(tenantId != null) {
-                String filterCondition = getMultiTenantFilterCondition(sessionFactoryBeanName, entity);
-                root.addFilter(GormProperties.TENANT_IDENTITY,filterCondition, true, Collections.<String, String>emptyMap(), Collections.<String, String>emptyMap());
-                mappings.addFilterDefinition(new FilterDefinition(
-                        GormProperties.TENANT_IDENTITY,
-                        filterCondition,
-                        Collections.singletonMap(GormProperties.TENANT_IDENTITY, root.getProperty(tenantId.getName()).getType())
-                ));
-            }
+        addMultiTenantFilterIfNecessary(entity, root, mappings, sessionFactoryBeanName);
 
-        }
         mappings.addEntityBinding(root);
     }
 
+    /**
+     * Add a Hibernate filter for multitenancy if the persistent class is multitenant
+     *
+     * @param entity target persistent entity for get tenant information
+     * @param persistentClass persistent class for add the filter and get tenant property info
+     * @param mappings mappings to add the filter
+     * @param sessionFactoryBeanName the session factory bean name
+     */
+    protected void addMultiTenantFilterIfNecessary(
+            HibernatePersistentEntity entity, PersistentClass persistentClass,
+            InFlightMetadataCollector mappings, String sessionFactoryBeanName) {
+        if (entity.isMultiTenant()) {
+            TenantId tenantId = entity.getTenantId();
+
+            if (tenantId != null) {
+                String filterCondition = getMultiTenantFilterCondition(sessionFactoryBeanName, entity);
+
+                persistentClass.addFilter(
+                        GormProperties.TENANT_IDENTITY,
+                        filterCondition,
+                        true,
+                        Collections.<String, String>emptyMap(),
+                        Collections.<String, String>emptyMap()
+                );
+
+                mappings.addFilterDefinition(new FilterDefinition(
+                        GormProperties.TENANT_IDENTITY,
+                        filterCondition,
+                        Collections.singletonMap(GormProperties.TENANT_IDENTITY, persistentClass.getProperty(tenantId.getName()).getType())
+                ));
+            }
+        }
+    }
 
     /**
      * Binds the sub classes of a root class using table-per-heirarchy inheritance mapping
@@ -1509,6 +1531,8 @@ public class GrailsDomainBinder implements MetadataContributor {
         else {
             bindSubClass(sub, subClass, mappings, sessionFactoryBeanName);
         }
+
+        addMultiTenantFilterIfNecessary(sub, subClass, mappings, sessionFactoryBeanName);
 
         final java.util.Collection<PersistentEntity> childEntities = sub.getMappingContext().getDirectChildEntities(sub);
         if (!childEntities.isEmpty()) {

